working/2509-private-named-parameters/feature-specification.md

```diff
@@ -268,6 +275,79 @@ class House({
 });
 ```
 
+### Concerns
+
+This proposal makes the language implicitly convert a private named parameter
+into the verbose pattern users do today where they declare a public named
+parameter and explicitly initialize the private field from it:
+
+```dart
+class C {
+  int? _variable;
+
+  C({int? variable}) : _variable = variable;
+}
+```
+
+While verbose, this code has the advantage of being very clear what's going on.
+A reader can see that the argument name they must use at the callsite is
+`variable`, the field is named `_variable`, and the latter is initialized from
+the former.
+
+Discarding the `_` implicitly may be confusing for users. If a user sees a class
+like:
+
+```dart
+class C({int? _variable}) {}
+```
+
+They may try to call it like `C(_variable: 123)` and be confused that it doesn't
+work. There's nothing in the code hinting that the `_` is removed from the
+argument name.
+
+In general, we try to design features that are both useful once you know
+them and intuitive to learn in the first place. This feature is helpful for
+brevity once you know the "trick", but it's opaque when it comes to learning.
+This is a real concern with the feature, but I believe the brevity makes it
+worth the learnability cost.
+
+We mitigate confusion here in a couple of ways:
+
+#### Only allow the syntax where it's meaningful
+
+At the language level, this proposal only allows `_` in a named parameter when
+doing so is *useful and meaningful*. It doesn't allow *any* named parameter to
+start with underscore, only a named parameter that declares or initializes a
+private instance field.
+
+A private named parameter *looks weird* since privacy makes little sense for an
+argument name and makes even less since for the local parameter variable. (We
+already have a lint that warns on using `_` for local variable names since it
+accomplishes nothing.)
+
+If a user sees `_` on a parameter and is trying to figure out what's going on,
+they will reliably be in a context where that parameter is also referring to a
+field. If we're lucky, that may lead them to intuit that the privacy is for the
+field, not the parameter.
+
+#### Provide a teaching errors if they use `_` for other named parameters
+
+If a user tries to put `_` before a named parameter that *isn't* an initializing
```

Comment:
FWIW, I don't expect the extra effort to implement this "teaching error" to carry its weight.  What is the user trying to accomplish by doing this?

Reply:
We have to report *some* error here because the proposal disallows it. I don't think it's any significant extra effort for that error to clarify that you can use `_` for constructor parameters that are initializing or declaring but not otherwise.

I agree it would be weird for users to try to do this, but my impression from reading lots of pub code over the years is that at scale, all possible user behavior will be observed no matter how irrational. :)